### PR TITLE
Add DateTime

### PR DESCRIPTION
**Pull request by albertkur**

### CODE REVIEW

File: src/date-time.ts [hunk 1]
```diff
@@ -0,0 +1,553 @@
...
+        if (zone.toLowerCase() === "utc")
+            return;
+
+        given(zone, "zone")
+            .ensureWhen(
+                zone.toLowerCase() === "local",
+                _ => false,
+                "should not use local zone")
+            .ensureWhen(
+                IANAZone.isValidSpecifier(zone),
```
Comment (shreymahendru) [inline src/date-time.ts:259]: Why do you have this? doesn't luxon validate this?

File: src/date-time.ts [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+                t => IANAZone.isValidZone(t),
+                "Invalid IANA zone")
+            .ensureWhen(
+                zone.toLowerCase().startsWith("utc+"),
+                t =>
+                {
+                    // range is +00:00 to +14:00
```
Comment (shreymahendru) [inline src/date-time.ts:266]: 14?? not 12?

File: src/date-time.ts [hunk 3]
```diff
@@ -0,0 +1,553 @@
...
+                    const offset = t.split("+").takeLast();
+
+                    if (offset.contains(":"))
```
Comment (shreymahendru) [inline src/date-time.ts:269]: You can normalize it before, rather than doing the `:` if/else
normalize as in it's what's passed is +11 => +11:00

File: src/date-time.ts [hunk 4]
```diff
@@ -0,0 +1,553 @@
...
+        return LuxonDateTime.now().setZone(zone).isValid;
+    }
+
+
+    private static _validateZone(zone: string): void
+    {
+        if (zone.toLowerCase() === "utc")
+            return;
+
+        given(zone, "zone")
```
Comment (shreymahendru) [inline src/date-time.ts:253]: make sure you `trim` the `zone`

File: src/date-time.ts [hunk 5]
```diff
@@ -0,0 +1,553 @@
...
+                },
+                "Invalid UTC offset for zone")
+            .ensureWhen(
+                zone.toLowerCase().startsWith("utc-"),
+                t =>
+                {
+                    // range is -00:00 to -12:00
+                    const offset = t.split("-").takeLast();
+
+                    if (offset.contains(":"))
```
Comment (shreymahendru) [inline src/date-time.ts:289]: same as above

File: src/date-time.ts [hunk 6]
```diff
@@ -0,0 +1,553 @@
...
+                "Invalid UTC offset for zone");
+    }
+
+
+    public override valueOf(): number
+    {
+        return this._dateTime.valueOf();
+    }
+
+    public equals(value: DateTime): boolean
```
Comment (shreymahendru) [inline src/date-time.ts:311]: equals should also take `null`

Comment (shreymahendru) [reply src/date-time.ts:311]: then the equals should first check null, 
then object identity (i.e reference check) 
then the values

File: src/date-time.ts [hunk 7]
```diff
@@ -0,0 +1,553 @@
...
+            zone: this._zone
+        });
+    }
+
+    public subtractTime(time: Duration): DateTime
+    {
+        given(time, "time").ensureHasValue().ensureIsObject();
+
+        return new DateTime({
+            value: this._dateTime.minus({ seconds: time.toSeconds() }).toFormat(DateTime._format),
```
Comment (shreymahendru) [inline src/date-time.ts:445]: same

File: test/date-time/date-time-properties.test.ts [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+    suite("Date value", () =>
+    {
+        test("check date code parsed correctly", () =>
+        {
+            Assert.strictEqual(new DateTime({ value: "2024-01-01 10:00", zone: "utc" }).dateValue, "2024-01-01");
+        });
+    });
+
+    suite("Time value", () =>
```
Comment (shreymahendru) [inline test/date-time/date-time-properties.test.ts:101]: Why we need these?

File: test/date-time/date-time-utility.test.ts [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    suite("Current zone", () =>
+    {
+        test("current zone is valid", () =>
+        {
+            Assert.ok(DateTime.validateTimeZone(DateTime.currentZone));
+        });
+
+        test("same as Luxon local zone", () =>
+        {
+            Assert.strictEqual(LuxonDateTime.now().zoneName, DateTime.currentZone);
```
Comment (shreymahendru) [inline test/date-time/date-time-utility.test.ts:17]: umm idk about this

**Review by shreymahendru — CHANGES_REQUESTED** [review]

File: src/date-time.ts [hunk 10]
```diff
@@ -0,0 +1,553 @@
...
+
+
+    private static _validateZone(zone: string): void
+    {
+        if (zone.toLowerCase() === "utc")
+            return;
+
+        given(zone, "zone")
+            .ensureWhen(
+                zone.toLowerCase() === "local",
```
[the remark this answers is not in the document]
Comment (albertkur) [reply src/date-time.ts:255]: luxon accepts "local" as a valid zone param and it will take the current system zone as the zone. since it won't be specific. i am just preventing use of local as a valid zone

File: src/date-time.ts (hunk 1 above)
Comment (albertkur) [reply src/date-time.ts:259]: I was being specific about the error

File: src/date-time.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (albertkur) [reply src/date-time.ts:255]: if we want to use local time zone we can use ` DateTime.currentZone`

File: src/date-time.ts (hunk 1 above)
Comment (shreymahendru) [reply src/date-time.ts:264]: Na don't need this

Comment (shreymahendru) [reply src/date-time.ts:264]: Imagine the list of all the zone, and it is probably iterating over all of them to find if it's in there or not. Now that is happening 2 times once here and once in creation of DateTime

File: src/date-time.ts [hunk 11]
```diff
@@ -0,0 +1,549 @@
...
+                zone.toLowerCase().startsWith("utc+"),
+                t =>
+                {
+                    // range is +00:00 to +14:00 (https://en.wikipedia.org/wiki/List_of_UTC_offsets)
+                    let offset = t.split("+").takeLast().trim();
+
+                    if (!offset.contains(":"))
+                        offset = `${offset}:00`;
+
+                    const [hour, minute] = offset.split(":").map(t => TypeHelper.parseNumber(t)!);
```
Comment (shreymahendru) [inline src/date-time.ts:277]: Parse number would return null if it's not a valid number. check that

File: test/date-time/date-time-create.test.ts [hunk 12]
```diff
@@ -0,0 +1,1068 @@
...
+
+            Assert.fail(`DateTime ${dateTime} is valid`);
+        }
+
+        test(`Given a valid value and zone
+        when the constructor is called 
+        then it should return a DateTime object`,
+            () =>
+            {
+                Assert.ok(new DateTime({ value: validValue, zone: "utc" }));
```
Comment (shreymahendru) [inline test/date-time/date-time-create.test.ts:34]: You can use ` Assert.doesNotThrow`

File: test/date-time/date-time-create.test.ts [hunk 13]
```diff
@@ -0,0 +1,1068 @@
...
+        {
+            let dateTime = "";
+            try
+            {
+                dateTime = new DateTime({ value, zone }).toString();
+            }
+            catch (e: any)
+            {
+                // console.log(e.reason);
+                Assert.ok(e.reason);
```
Comment (shreymahendru) [inline test/date-time/date-time-create.test.ts:22]: umm not sure about this, Asset.ok take in a boolean, what if the reason was an empty string it would fail this

File: test/date-time/date-time-create.test.ts [hunk 14]
```diff
@@ -0,0 +1,1068 @@
...
+                Assert.ok(new DateTime({ value: validValue, zone: "utc" }));
+            }
+        );
+
+        test(`Given value as empty string and a valid zone
+        when the constructor is called 
+        then it should throw a validation error`,
+            () =>
+            {
+                checkIsInvalid("");
```
Comment (shreymahendru) [inline test/date-time/date-time-create.test.ts:43]: You should do this test with Assert.throws, there you can also check the error thrown is the error exepcted

File: test/date-time/date-time-create.test.ts (hunk 12 above)
Comment (shreymahendru) [reply test/date-time/date-time-create.test.ts:34]: for this, if the test fails it won't really show that it failed it will blow up with an exception

File: test/date-time/date-time-create.test.ts [hunk 15]
```diff
@@ -0,0 +1,1068 @@
...
+
+    suite("Now", () =>
+    {
+        test(`Given DateTime.now() in default zone (utc)
+        when it's compared to Luxon DateTime now in utc
+        then it should be same`,
+            () =>
+            {
+                // this might fail if at the end of minute and next is at start of minute. but rare condition
+                Assert.strictEqual(DateTime.now().value, LuxonDateTime.utc().toFormat("yyyy-MM-dd HH:mm"));
```
Comment (shreymahendru) [inline test/date-time/date-time-create.test.ts:346]: yeeea i don't think you need to test this

Comment (shreymahendru) [reply test/date-time/date-time-create.test.ts:346]: unless you want to temporarily replave Date.now to return what you want it to return. it so it always exact

Comment (shreymahendru) [reply test/date-time/date-time-create.test.ts:346]: lets talk about these

File: test/date-time/date-time-create.test.ts [hunk 16]
```diff
@@ -0,0 +1,1068 @@
...
+        test(`Given zone as America/Los_Angeles
+        when DateTime for now is created 
+        then zone property should be America/Los_Angeles`,
+            () =>
+            {
+                Assert.strictEqual(DateTime.now("America/Los_Angeles").zone, "America/Los_Angeles");
+            }
+        );
+
+        // zone validation is done in constructor
```
Comment (shreymahendru) [inline test/date-time/date-time-create.test.ts:407]: test shouldn't need to know what the implimentation details

File: test/date-time/date-time-create.test.ts [hunk 17]
```diff
@@ -0,0 +1,1068 @@
...
+        test(`Given a valid time code(1000) and zone, and date code as an empty string
+        when DateTime is created from that codes
+        then  it should throw a validation error`,
+            () =>
+            {
+                checkIsInvalid("", validTimeCode);
+            }
+        );
+
+        test(`Given a valid time code(1000) and zone, and date code contain special characters 2024-01-01
```
Comment (shreymahendru) [inline test/date-time/date-time-create.test.ts:579]: it's just an invalid format. why say special characters in the description?

File: test/date-time/date-time-create.test.ts [hunk 18]
```diff
@@ -0,0 +1,1068 @@
...
+        test(`Given a valid date code(20240101) and zone, and time code with invalid hour 2400
+        when DateTime is created from that codes
+        then  it should throw a validation error`,
+            () =>
+            {
+                checkIsInvalid(validDateCode, "2400");
+            }
+        );
+
+        test(`Given a valid time code(1000) and zone, and date code with invalid day 0 20240100
```
Comment (shreymahendru) [inline test/date-time/date-time-create.test.ts:687]: move all the datecode stuff together and timecode together

File: test/date-time/date-time-create.test.ts [hunk 19]
```diff
@@ -0,0 +1,1068 @@
...
+        test(`Given a valid time code(1000) and zone, and date code with invalid day feb 30 20240230
+        when DateTime is created from that codes
+        then  it should throw a validation error`,
+            () =>
+            {
+                checkIsInvalid("20240230", validTimeCode);
+            }
+        );
+
+        test(`Given a valid time code(1000) and zone, and date code feb 29 on leap year 20240229
```
Comment (shreymahendru) [inline test/date-time/date-time-create.test.ts:723]: I don't think you need to check these. since this is checked in creation. and you have already tested that

Comment (shreymahendru) [reply test/date-time/date-time-create.test.ts:723]: All you can check here would just be the format of the datecode and timecode

File: test/date-time/date-time-create.test.ts [hunk 20]
```diff
@@ -0,0 +1,1068 @@
...
+        test(`Given a valid date code(20240101), time code(1000) and zone as UTC+5:30
+        when DateTime is created from that codes
+        then zone property should be UTC+5:30`,
+            () =>
+            {
+                Assert.strictEqual(DateTime.createFromCodes(validDateCode, validTimeCode, "UTC+5:30").zone, "UTC+5:30");
+            }
+        );
+
+        test(`Given a valid date code(20240101), time code(1000) and zone as America/Los_Angeles
```
Comment (shreymahendru) [inline test/date-time/date-time-create.test.ts:786]: Add test for the happy path, given this datecode and timecode values the expected datetime value is created

File: test/date-time/date-time-create.test.ts [hunk 21]
```diff
@@ -0,0 +1,1068 @@
...
+        test(`Given a valid time value(10:00) and zone, and date value with invalid day 32 2024-01-32
+        when DateTime is created from that values
+        then  it should throw a validation error`,
+            () =>
+            {
+                checkIsInvalid("2024-01-32", validTimeValue);
+            }
+        );
+
+        test(`Given a valid time value(10:00) and zone, and date value with invalid day feb 29 on non-leap year 2023-02-29
```
Comment (shreymahendru) [inline test/date-time/date-time-create.test.ts:975]: You don't really need these one

File: test/date-time/date-time-properties.test.ts [hunk 22]
```diff
@@ -0,0 +1,259 @@
+import * as Assert from "assert";
+import { DateTime } from "../../src/date-time";
+import { DateTime as LuxonDateTime } from "luxon";
+import { Duration } from "../../src";
+
+
+suite("DateTime Properties", () =>
+{
+    suite("Value", () =>
```
Comment (shreymahendru) [inline test/date-time/date-time-properties.test.ts:9]: do you really need to test this. this will be tested implicitly by other tests

File: test/date-time/date-time-properties.test.ts [hunk 23]
```diff
@@ -0,0 +1,259 @@
...
+            () =>
+            {
+                const dateTime = new DateTime({ value: validDateTime, zone: "utc" });
+                Assert.ok(DateTime.validateDateTimeFormat(dateTime.value));
+            }
+        );
+    });
+
+
+    suite("Zone", () =>
```
Comment (shreymahendru) [inline test/date-time/date-time-properties.test.ts:34]: same for this

File: test/date-time/date-time-properties.test.ts [hunk 24]
```diff
@@ -0,0 +1,259 @@
...
+
+
+    suite("Timestamp", () =>
+    {
+        test(`Given a luxon date time
+        when DateTime is created from that value and zone
+        then both should have same timeStamp`,
+            () =>
+            {
+                const timeStamp = LuxonDateTime.fromFormat("2024-01-01 10:00", "yyyy-MM-dd HH:mm", { zone: "utc" }).toSeconds();
```
Comment (shreymahendru) [inline test/date-time/date-time-properties.test.ts:74]: you should pre-calculate the timestamp and test against than rather than use Luxon here

Comment (shreymahendru) [reply test/date-time/date-time-properties.test.ts:74]: like you did in the test below

File: test/date-time/date-time-properties.test.ts [hunk 25]
```diff
@@ -0,0 +1,259 @@
...
+                Assert.strictEqual(dateTime.dateValue, dateValue);
+            }
+        );
+
+        test(`Given a valid value (${value}) 
+        when a DateTime is created from that value
+        then the date value property should be in valid date format`,
+            () =>
+            {
+                Assert.ok(DateTime.validateDateFormat(dateTime.dateValue));
```
Comment (shreymahendru) [inline test/date-time/date-time-properties.test.ts:151]: you are validating one functionality with other functionality of the same thing, this test will fail if either of the 2 are messed up

Comment (shreymahendru) [reply test/date-time/date-time-properties.test.ts:151]: you can use `matchesFormat` to validate format

File: test/date-time/date-time-properties.test.ts [hunk 26]
```diff
@@ -0,0 +1,259 @@
...
+                Assert.strictEqual(dateTime.timeValue, timeValue);
+            }
+        );
+
+        test(`Given a valid value (${value}) 
+        when a DateTime is created from that value
+        then the time value property should be in valid time format`,
+            () =>
+            {
+                Assert.ok(DateTime.validateTimeFormat(dateTime.timeValue));
```
Comment (shreymahendru) [inline test/date-time/date-time-properties.test.ts:176]: same here

File: test/date-time/date-time-properties.test.ts (hunk 25 above)
Comment (shreymahendru) [reply test/date-time/date-time-properties.test.ts:151]: or you can just check you get the expected value from dateTime.dateValue

File: test/date-time/date-time-utility.test.ts [hunk 27]
```diff
@@ -0,0 +1,397 @@
...
+        const zone = "utc";
+        const dateTime = new DateTime({ value: value, zone: zone });
+        const luxonDateTime = LuxonDateTime.fromFormat(value, "yyyy-MM-dd HH:mm", { zone });
+
+        test(`Given a DateTime with value "2024-01-01 10:00" and zone utc
+        when it's compared to luxon dateTime representing the same 
+        then it's valueOf() should be same as Luxon DateTime`,
+            () =>
+            {
+                Assert.strictEqual(dateTime.valueOf(), luxonDateTime.valueOf());
```
Comment (shreymahendru) [inline test/date-time/date-time-utility.test.ts:41]: try not to use Luxon here

File: test/date-time/date-time-utility.test.ts [hunk 28]
```diff
@@ -0,0 +1,397 @@
...
+        test(`Given a DateTime with value "2024-01-01 10:00" and zone utc
+        when it's valueOf() is validated with luxon
+        then it should return true`,
+            () =>
+            {
+                Assert.ok(LuxonDateTime.fromMillis(dateTime.valueOf()).isValid);
+            }
+        );
+
+        test(`Given a DateTime with value "2024-01-01 10:00" and zone utc
```
Comment (shreymahendru) [inline test/date-time/date-time-utility.test.ts:54]: What are you testing here?

File: test/date-time/date-time-utility.test.ts [hunk 29]
```diff
@@ -0,0 +1,397 @@
...
+                Assert.strictEqual(new DateTime({ value, zone: "utc" }).toStringDateTime(), value);
+            }
+        );
+
+        test(`Given a valid value (${value}) and zone (utc)
+        when a DateTime is created from that value and zone
+        then toStringDateTime() should return the value in a valid DateTime format`,
+            () =>
+            {
+                Assert.ok(DateTime.validateDateTimeFormat(new DateTime({ value: value, zone: "utc" }).toStringDateTime()));
```
Comment (shreymahendru) [inline test/date-time/date-time-utility.test.ts:148]: ?

Comment (shreymahendru) [reply test/date-time/date-time-utility.test.ts:148]: for toString have a few examples where you know the output and just check against it

File: test/date-time/date-time-utility.test.ts [hunk 30]
```diff
@@ -0,0 +1,397 @@
...
+    suite("to string ISO", () =>
+    {
+        const value = "2024-01-01 10:00";
+
+        test(`Given a valid value (${value}) and zone (utc)
+        when a DateTime is created from that value and zone
+        then toStringISO() should return a valid ISO string`,
+            () =>
+            {
+                Assert.ok(LuxonDateTime.fromISO(new DateTime({ value, zone: "utc" }).toStringISO()).isValid);
```
Comment (shreymahendru) [inline test/date-time/date-time-utility.test.ts:173]: same here.

File: test/date-time/date-time-utility.test.ts [hunk 31]
```diff
@@ -0,0 +1,397 @@
...
+        });
+    });
+
+    suite("Convert to zone", () =>
+    {
+        const value = "2024-01-01 10:00";
+
+        const dateTime = new DateTime({ value, zone: "utc" });
+
+        test(`Given a DateTime object with value (${value}) and zone (utc)
```
Comment (shreymahendru) [inline test/date-time/date-time-utility.test.ts:336]: Have zone coversion test for the Daylist saving times is on vs not on for a particular zone

File: test/date-time/date-time-math.test.ts [hunk 32]
```diff
@@ -0,0 +1,281 @@
...
+            given(time, "time").ensureHasValue().ensureIsString();
+            given(expectedValue, "expectedValue").ensureHasValue().ensureIsString()
+                .ensure(t => DateTime.validateDateTimeFormat(t));
+
+            test(`Given a DateTime (${dateTime.toString()}), and a duration of ${time}
+            when duration is added to the dateTime
+            then it should return value of ${expectedValue}`,
+                () =>
+                {
+                    Assert.strictEqual(dateTime.addTime(duration).value, expectedValue);
```
Comment (shreymahendru) [inline test/date-time/date-time-math.test.ts:27]: You should check the return of valueOf

Comment (shreymahendru) [reply test/date-time/date-time-math.test.ts:27]: that way you can check milli second math

File: test/date-time/date-time-math.test.ts [hunk 33]
```diff
@@ -0,0 +1,281 @@
...
+            given(time, "time").ensureHasValue().ensureIsString();
+            given(expectedValue, "expectedValue").ensureHasValue().ensureIsString()
+                .ensure(t => DateTime.validateDateTimeFormat(t));
+
+            test(`Given a DateTime (${dateTime.toString()}), and a duration of ${time}
+            when duration is subtracted from the dateTime
+            then it should return a value of ${expectedValue}`,
+                () =>
+                {
+                    Assert.strictEqual(dateTime.subtractTime(duration).value, expectedValue);
```
Comment (shreymahendru) [inline test/date-time/date-time-math.test.ts:98]: same here

File: test/date-time/date-time-math.test.ts [hunk 34]
```diff
@@ -0,0 +1,281 @@
...
+                given(days, "days").ensureHasValue().ensureIsNumber();
+                given(expectedValue, "expectedValue").ensureHasValue().ensureIsString()
+                    .ensure(t => DateTime.validateDateTimeFormat(t));
+
+                test(`Given a DateTime (${dateTime.toString()}), and number of days ${days}
+                when number of days is added to the dateTime
+                then it should return value of ${expectedValue}`,
+                    () =>
+                    {
+                        Assert.strictEqual(dateTime.addDays(days).value, expectedValue);
```
Comment (shreymahendru) [inline test/date-time/date-time-math.test.ts:173]: here checking with string is fine

File: test/date-time/date-time-math.test.ts [hunk 35]
```diff
@@ -0,0 +1,281 @@
...
+                    }
+                );
+            }
+
+            checkIsCorrect(dateTime, 0, dateTime.value);
+            checkIsCorrect(dateTime, 1, "2024-01-02 10:00");
+            checkIsCorrect(dateTime, 31, "2024-02-01 10:00");
+            checkIsCorrect(dateTime, 366, "2025-01-01 10:00");
```
Comment (shreymahendru) [inline test/date-time/date-time-math.test.ts:181]: Can you check what happens when you add a day and that's the day DST start/ends does the time change?

File: test/date-time/date-time-math.test.ts [hunk 36]
```diff
@@ -0,0 +1,281 @@
...
+                    () =>
+                    {
+                        try
+                        {
+                            dateTime.addDays(days);
+                        }
+                        catch (e: any)
+                        {
+                            // console.log(e.reason);
+                            Assert.ok(e.reason);
```
Comment (shreymahendru) [inline test/date-time/date-time-math.test.ts:201]: use assert throws

File: test/date-time/date-time-comparison.test.ts [hunk 37]
```diff
@@ -0,0 +1,1109 @@
+import * as Assert from "assert";
+import { DateTime } from "../../src/date-time";
+import { Duration } from "../../src";
+import { given } from "@nivinjoseph/n-defensive";
+import { IANAZone } from "luxon";
+
+
+suite("DateTime Comparison", () =>
```
Comment (shreymahendru) [inline test/date-time/date-time-comparison.test.ts:8]: This suite should be structured by method names

and then in there have these has cases. 
same date 
positive case 
negative case
edge cases

File: test/date-time/date-time-comparison.test.ts [hunk 38]
```diff
@@ -0,0 +1,1109 @@
...
+            when checking the difference between them in calendar days
+            then it should return 0`,
+                () =>
+                {
+                    Assert.strictEqual(dateTime.daysDiff(dateTime), 0);
+                }
+            );
+        });
+
+        suite("Compare same value and zone", () =>
```
Comment (shreymahendru) [inline test/date-time/date-time-comparison.test.ts:116]: Why you testing this. isn't it the same as above?

File: test/date-time/date-time-utility.test.ts (hunk 28 above)
Comment (albertkur) [reply test/date-time/date-time-utility.test.ts:54]: valueOf( should return milliSeconds since epoch which should be timestamp *1000

File: test/date-time/date-time-utility.test.ts (hunk 31 above)
Comment (albertkur) [reply test/date-time/date-time-utility.test.ts:336]: ?

**Review by nivinjoseph — APPROVED** [review]
